// websocket-driver/lib/websocket/driver/base.js
var Emitter = require('events').EventEmitter,
    util    = require('util'),
    streams = require('../streams'),
    Headers = require('./headers');

var Base = function(request, url, options) {
  Emitter.call(this);

  this._request   = request;
  this._options   = options || {***REMOVED***;
  this._maxLength = this._options.maxLength || this.MAX_LENGTH;
  this.__headers  = new Headers();
  this.__queue    = [];
  this.readyState = 0;
  this.url        = url;

  this.io = new streams.IO(this);
  this.messages = new streams.Messages(this);
  this._bindEventListeners();
***REMOVED***;
util.inherits(Base, Emitter);

var instance = {
  // This is 64MB, small enough for an average VPS to handle without
  // crashing from process out of memory
  MAX_LENGTH: 0x3ffffff,

  STATES: ['connecting', 'open', 'closing', 'closed'],

  _bindEventListeners: function() {
    var self = this;

    // Protocol errors are informational and do not have to be handled
    this.messages.on('error', function() {***REMOVED***);

    this.on('message', function(event) {
      var messages = self.messages;
      if (messages.readable) messages.emit('data', event.data);
    ***REMOVED***);

    this.on('error', function(error) {
      var messages = self.messages;
      if (messages.readable) messages.emit('error', error);
    ***REMOVED***);

    this.on('close', function() {
      var messages = self.messages;
      if (!messages.readable) return;
      messages.readable = messages.writable = false;
      messages.emit('end');
    ***REMOVED***);
  ***REMOVED***,

  getState: function() {
    return this.STATES[this.readyState] || null;
  ***REMOVED***,

  setHeader: function(name, value) {
    if (this.readyState > 0) return false;
    this.__headers.set(name, value);
    return true;
  ***REMOVED***,

  start: function() {
    if (this.readyState !== 0) return false;
    this._write(this._handshakeResponse());
    if (this._stage !== -1) this._open();
    return true;
  ***REMOVED***,

  text: function(message) {
    return this.frame(message);
  ***REMOVED***,

  binary: function(message) {
    return false;
  ***REMOVED***,

  ping: function() {
    return false;
  ***REMOVED***,

  close: function(reason, code) {
    if (this.readyState !== 1) return false;
    this.readyState = 3;
    this.emit('close', new Base.CloseEvent(null, null));
    return true;
  ***REMOVED***,

  _open: function() {
    this.readyState = 1;
    this.__queue.forEach(function(args) { this.frame.apply(this, args) ***REMOVED***, this);
    this.__queue = [];
    this.emit('open', new Base.OpenEvent());
  ***REMOVED***,

  _queue: function(message) {
    this.__queue.push(message);
    return true;
  ***REMOVED***,

  _write: function(chunk) {
    var io = this.io;
    if (io.readable) io.emit('data', chunk);
  ***REMOVED***
***REMOVED***;

for (var key in instance)
  Base.prototype[key] = instance[key];


Base.ConnectEvent = function() {***REMOVED***;

Base.OpenEvent = function() {***REMOVED***;

Base.CloseEvent = function(code, reason) {
  this.code   = code;
  this.reason = reason;
***REMOVED***;

Base.MessageEvent = function(data) {
  this.data = data;
***REMOVED***;

module.exports = Base;
