// sockjs/lib/chunking-test.js
(function() {
  var utils;

  utils = require('./utils');

  exports.app = {
    chunking_test: function(req, res, _, next_filter) {
      var write,
        _this = this;
      res.setHeader('Content-Type', 'application/javascript; charset=UTF-8');
      res.writeHead(200);
      write = function(payload) {
        try {
          return res.write(payload + '\n');
        ***REMOVED*** catch (x) {

        ***REMOVED***
      ***REMOVED***;
      utils.timeout_chain([
        [
          0, function() {
            return write('h');
          ***REMOVED***
        ], [
          1, function() {
            return write(Array(2049).join(' ') + 'h');
          ***REMOVED***
        ], [
          5, function() {
            return write('h');
          ***REMOVED***
        ], [
          25, function() {
            return write('h');
          ***REMOVED***
        ], [
          125, function() {
            return write('h');
          ***REMOVED***
        ], [
          625, function() {
            return write('h');
          ***REMOVED***
        ], [
          3125, function() {
            write('h');
            return res.end();
          ***REMOVED***
        ]
      ]);
      return true;
    ***REMOVED***,
    info: function(req, res, _) {
      var info;
      info = {
        websocket: this.options.websocket,
        origins: ['*:*'],
        cookie_needed: !!this.options.jsessionid,
        entropy: utils.random32()
      ***REMOVED***;
      if (typeof this.options.base_url === 'function') {
        info.base_url = this.options.base_url();
      ***REMOVED*** else if (this.options.base_url) {
        info.base_url = this.options.base_url;
      ***REMOVED***
      res.setHeader('Content-Type', 'application/json; charset=UTF-8');
      res.writeHead(200);
      return res.end(JSON.stringify(info));
    ***REMOVED***,
    info_options: function(req, res) {
      res.statusCode = 204;
      res.setHeader('Access-Control-Allow-Methods', 'OPTIONS, GET');
      res.setHeader('Access-Control-Max-Age', res.cache_for);
      return '';
    ***REMOVED***
  ***REMOVED***;

***REMOVED***).call(this);
