// sockjs/lib/transport.js
(function() {
  var GenericReceiver, MAP, ResponseReceiver, Session, SockJSConnection, Transport, closeFrame, register, stream, utils, uuid,
    __hasProp = Object.prototype.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; ***REMOVED*** function ctor() { this.constructor = child; ***REMOVED*** ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; ***REMOVED***;

  stream = require('stream');

  uuid = require('node-uuid');

  utils = require('./utils');

  Transport = (function() {

    function Transport() {***REMOVED***

    return Transport;

  ***REMOVED***)();

  Transport.CONNECTING = 0;

  Transport.OPEN = 1;

  Transport.CLOSING = 2;

  Transport.CLOSED = 3;

  closeFrame = function(status, reason) {
    return 'c' + JSON.stringify([status, reason]);
  ***REMOVED***;

  SockJSConnection = (function(_super) {

    __extends(SockJSConnection, _super);

    function SockJSConnection(_session) {
      this._session = _session;
      this.id = uuid();
      this.headers = {***REMOVED***;
      this.prefix = this._session.prefix;
    ***REMOVED***

    SockJSConnection.prototype.toString = function() {
      return '<SockJSConnection ' + this.id + '>';
    ***REMOVED***;

    SockJSConnection.prototype.write = function(string) {
      return this._session.send('' + string);
    ***REMOVED***;

    SockJSConnection.prototype.end = function(string) {
      if (string) this.write(string);
      this.close();
      return null;
    ***REMOVED***;

    SockJSConnection.prototype.close = function(code, reason) {
      return this._session.close(code, reason);
    ***REMOVED***;

    SockJSConnection.prototype.destroy = function() {
      this.removeAllListeners();
      return this.end();
    ***REMOVED***;

    SockJSConnection.prototype.destroySoon = function() {
      return this.destroy();
    ***REMOVED***;

    return SockJSConnection;

  ***REMOVED***)(stream.Stream);

  SockJSConnection.prototype.__defineGetter__('readable', function() {
    return this._session.readyState === Transport.OPEN;
  ***REMOVED***);

  SockJSConnection.prototype.__defineGetter__('writable', function() {
    return this._session.readyState === Transport.OPEN;
  ***REMOVED***);

  SockJSConnection.prototype.__defineGetter__('readyState', function() {
    return this._session.readyState;
  ***REMOVED***);

  MAP = {***REMOVED***;

  Session = (function() {

    function Session(session_id, server) {
      var _this = this;
      this.session_id = session_id;
      this.heartbeat_delay = server.options.heartbeat_delay;
      this.disconnect_delay = server.options.disconnect_delay;
      this.prefix = server.options.prefix;
      this.send_buffer = [];
      this.is_closing = false;
      this.readyState = Transport.CONNECTING;
      if (this.session_id) MAP[this.session_id] = this;
      this.timeout_cb = function() {
        return _this.didTimeout();
      ***REMOVED***;
      this.to_tref = setTimeout(this.timeout_cb, this.disconnect_delay);
      this.connection = new SockJSConnection(this);
      this.emit_open = function() {
        _this.emit_open = null;
        return server.emit('connection', _this.connection);
      ***REMOVED***;
    ***REMOVED***

    Session.prototype.register = function(req, recv) {
      if (this.recv) {
        recv.doSendFrame(closeFrame(2010, "Another connection still open"));
        recv.didClose();
        return;
      ***REMOVED***
      if (this.to_tref) {
        clearTimeout(this.to_tref);
        this.to_tref = null;
      ***REMOVED***
      if (this.readyState === Transport.CLOSING) {
        this.flushToRecv(recv);
        recv.doSendFrame(this.close_frame);
        recv.didClose();
        this.to_tref = setTimeout(this.timeout_cb, this.disconnect_delay);
        return;
      ***REMOVED***
      this.recv = recv;
      this.recv.session = this;
      this.decorateConnection(req);
      if (this.readyState === Transport.CONNECTING) {
        this.recv.doSendFrame('o');
        this.readyState = Transport.OPEN;
        process.nextTick(this.emit_open);
      ***REMOVED***
      if (!this.recv) return;
      this.tryFlush();
    ***REMOVED***;

    Session.prototype.decorateConnection = function(req) {
      var address, headers, key, remoteAddress, remotePort, socket, _i, _len, _ref;
      if (!(socket = this.recv.connection)) socket = this.recv.response.connection;
      try {
        remoteAddress = socket.remoteAddress;
        remotePort = socket.remotePort;
        address = socket.address();
      ***REMOVED*** catch (x) {

      ***REMOVED***
      if (remoteAddress) {
        this.connection.remoteAddress = remoteAddress;
        this.connection.remotePort = remotePort;
        this.connection.address = address;
      ***REMOVED***
      this.connection.url = req.url;
      this.connection.pathname = req.pathname;
      this.connection.protocol = this.recv.protocol;
      headers = {***REMOVED***;
      _ref = ['referer', 'x-client-ip', 'x-forwarded-for', 'x-cluster-client-ip', 'via', 'x-real-ip', 'host', 'user-agent', 'accept-language'];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        key = _ref[_i];
        if (req.headers[key]) headers[key] = req.headers[key];
      ***REMOVED***
      if (headers) return this.connection.headers = headers;
    ***REMOVED***;

    Session.prototype.unregister = function() {
      this.recv.session = null;
      this.recv = null;
      if (this.to_tref) clearTimeout(this.to_tref);
      return this.to_tref = setTimeout(this.timeout_cb, this.disconnect_delay);
    ***REMOVED***;

    Session.prototype.flushToRecv = function(recv) {
      var sb, _ref;
      if (this.send_buffer.length > 0) {
        _ref = [this.send_buffer, []], sb = _ref[0], this.send_buffer = _ref[1];
        recv.doSendBulk(sb);
        return true;
      ***REMOVED***
      return false;
    ***REMOVED***;

    Session.prototype.tryFlush = function() {
      var x,
        _this = this;
      if (!this.flushToRecv(this.recv)) {
        if (this.to_tref) clearTimeout(this.to_tref);
        x = function() {
          if (_this.recv) {
            _this.to_tref = setTimeout(x, _this.heartbeat_delay);
            return _this.recv.doSendFrame("h");
          ***REMOVED***
        ***REMOVED***;
        this.to_tref = setTimeout(x, this.heartbeat_delay);
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.didTimeout = function() {
      if (this.to_tref) {
        clearTimeout(this.to_tref);
        this.to_tref = null;
      ***REMOVED***
      if (this.readyState !== Transport.CONNECTING && this.readyState !== Transport.OPEN && this.readyState !== Transport.CLOSING) {
        throw Error('INVALID_STATE_ERR');
      ***REMOVED***
      if (this.recv) throw Error('RECV_STILL_THERE');
      this.readyState = Transport.CLOSED;
      this.connection.emit('end');
      this.connection.emit('close');
      this.connection = null;
      if (this.session_id) {
        delete MAP[this.session_id];
        return this.session_id = null;
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.didMessage = function(payload) {
      if (this.readyState === Transport.OPEN) {
        this.connection.emit('data', payload);
      ***REMOVED***
    ***REMOVED***;

    Session.prototype.send = function(payload) {
      if (this.readyState !== Transport.OPEN) return false;
      this.send_buffer.push('' + payload);
      if (this.recv) this.tryFlush();
      return true;
    ***REMOVED***;

    Session.prototype.close = function(status, reason) {
      if (status == null) status = 1000;
      if (reason == null) reason = "Normal closure";
      if (this.readyState !== Transport.OPEN) return false;
      this.readyState = Transport.CLOSING;
      this.close_frame = closeFrame(status, reason);
      if (this.recv) {
        this.recv.doSendFrame(this.close_frame);
        if (this.recv) this.recv.didClose();
        if (this.recv) this.unregister();
      ***REMOVED***
      return true;
    ***REMOVED***;

    return Session;

  ***REMOVED***)();

  Session.bySessionId = function(session_id) {
    if (!session_id) return null;
    return MAP[session_id] || null;
  ***REMOVED***;

  register = function(req, server, session_id, receiver) {
    var session;
    session = Session.bySessionId(session_id);
    if (!session) session = new Session(session_id, server);
    session.register(req, receiver);
    return session;
  ***REMOVED***;

  exports.register = function(req, server, receiver) {
    return register(req, server, req.session, receiver);
  ***REMOVED***;

  exports.registerNoSession = function(req, server, receiver) {
    return register(req, server, void 0, receiver);
  ***REMOVED***;

  GenericReceiver = (function() {

    function GenericReceiver(thingy) {
      this.thingy = thingy;
      this.setUp(this.thingy);
    ***REMOVED***

    GenericReceiver.prototype.setUp = function() {
      var _this = this;
      this.thingy_end_cb = function() {
        return _this.didAbort(1006, "Connection closed");
      ***REMOVED***;
      this.thingy.addListener('close', this.thingy_end_cb);
      return this.thingy.addListener('end', this.thingy_end_cb);
    ***REMOVED***;

    GenericReceiver.prototype.tearDown = function() {
      this.thingy.removeListener('close', this.thingy_end_cb);
      this.thingy.removeListener('end', this.thingy_end_cb);
      return this.thingy_end_cb = null;
    ***REMOVED***;

    GenericReceiver.prototype.didAbort = function(status, reason) {
      var session;
      session = this.session;
      this.didClose(status, reason);
      if (session) return session.didTimeout();
    ***REMOVED***;

    GenericReceiver.prototype.didClose = function(status, reason) {
      if (this.thingy) {
        this.tearDown(this.thingy);
        this.thingy = null;
      ***REMOVED***
      if (this.session) return this.session.unregister(status, reason);
    ***REMOVED***;

    GenericReceiver.prototype.doSendBulk = function(messages) {
      var m, q_msgs;
      q_msgs = (function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = messages.length; _i < _len; _i++) {
          m = messages[_i];
          _results.push(utils.quote(m));
        ***REMOVED***
        return _results;
      ***REMOVED***)();
      return this.doSendFrame('a' + '[' + q_msgs.join(',') + ']');
    ***REMOVED***;

    return GenericReceiver;

  ***REMOVED***)();

  ResponseReceiver = (function(_super) {

    __extends(ResponseReceiver, _super);

    ResponseReceiver.prototype.max_response_size = void 0;

    function ResponseReceiver(request, response, options) {
      this.request = request;
      this.response = response;
      this.options = options;
      this.curr_response_size = 0;
      try {
        this.request.connection.setKeepAlive(true, 5000);
      ***REMOVED*** catch (x) {

      ***REMOVED***
      ResponseReceiver.__super__.constructor.call(this, this.request.connection);
      if (this.max_response_size === void 0) {
        this.max_response_size = this.options.response_limit;
      ***REMOVED***
    ***REMOVED***

    ResponseReceiver.prototype.doSendFrame = function(payload) {
      var r;
      this.curr_response_size += payload.length;
      r = false;
      try {
        this.response.write(payload);
        r = true;
      ***REMOVED*** catch (x) {

      ***REMOVED***
      if (this.max_response_size && this.curr_response_size >= this.max_response_size) {
        this.didClose();
      ***REMOVED***
      return r;
    ***REMOVED***;

    ResponseReceiver.prototype.didClose = function() {
      ResponseReceiver.__super__.didClose.apply(this, arguments);
      try {
        this.response.end();
      ***REMOVED*** catch (x) {

      ***REMOVED***
      return this.response = null;
    ***REMOVED***;

    return ResponseReceiver;

  ***REMOVED***)(GenericReceiver);

  exports.GenericReceiver = GenericReceiver;

  exports.Transport = Transport;

  exports.Session = Session;

  exports.ResponseReceiver = ResponseReceiver;

  exports.SockJSConnection = SockJSConnection;

***REMOVED***).call(this);
